File: src/Strategy.py

from Exchange import Exchange
from Broker import Broker
from Order import *

class Strategy():
    def __init__(self, exchange : Exchange, broker : Broker) -> None:
        self.exchange = exchange
        self.broker = broker

    def reset(self):
        self.exchange.reset()
        self.broker.rest()

    def next(self):
        return []

class BenchMarkBH(Strategy):
    def __init__(self, exchange: Exchange, asset_name : str) -> None:
        self.broker = Broker(exchange=exchange)
        self.broker.cheat_on_close = True
        super().__init__(exchange, self.broker)
        self.invested = False
        self.asset_name = asset_name

    def next(self):
        if not self.invested:
            new_order = MarketOrder(
                order_create_time = self.exchange.market_time,
                asset_name = self.asset_name,
                units = self.broker.cash / (self.exchange.market_view[self.asset_name]["CLOSE"])
            )
            self.invested = True
            return [new_order]

File: src/c++/Strategy.cpp
#include "pch.h"
#include "Strategy.h"
#include "Exchange.h"
#include "Broker.h"
#include "Order.h"

Strategy::Strategy(__Exchange &exchangeObj, __Broker &brokerObj) :
	__exchange(exchangeObj), __broker(brokerObj)
{
}
void Strategy::next() {
}
void TestStrategy::register_test_map(std::vector<order_schedule>orders) {
	this->order_scheduler = orders;
}

void TestStrategy::next() {
	for (auto it = this->order_scheduler.begin(); it != this->order_scheduler.end();) {
		if (it->i == i) {
			switch (it->order_type) {
			case MARKET_ORDER: {
				this->__broker._place_market_order(it->asset_id, it->units);
				break;
			}
			case LIMIT_ORDER: {
				this->__broker._place_limit_order(it->asset_id, it->units, it->limit);
				break;
			}
			case STOP_LOSS_ORDER: {
				Position* existing_position = &this->__broker.portfolio[it->asset_id];
				this->__broker.place_stoploss_order(existing_position, it->units, it->limit);
				break;
			}
			}
			it = this->order_scheduler.erase(it);
		}
		else {
			it++;
		}
	}
	i++;
}